tsl/test/isolation/specs/continuous_aggs_concurrent_refresh.spec

```diff
@@ -65,111 +74,220 @@ setup
+# buckets/rows.
 session "S1"
 setup
 {
-    SET LOCAL lock_timeout = '500ms';
-    SET LOCAL deadlock_timeout = '100ms';
+    SET SESSION lock_timeout = '500ms';
+    SET SESSION deadlock_timeout = '500ms';
 }
 step "S1_select"
 {
-    SELECT day, avg_temp
-    FROM daily_temp
+    SELECT bucket, avg_temp
+    FROM cond_10
     ORDER BY 1;
 
-    SELECT * FROM cagg_bucket_count('daily_temp');
+    SELECT * FROM cagg_bucket_count('cond_10');
+    SELECT h.table_name AS hypertable, it.watermark AS threshold
+    FROM _timescaledb_catalog.continuous_aggs_invalidation_threshold it,
+    _timescaledb_catalog.hypertable h
+    WHERE it.hypertable_id = h.id;
 }
 
-# Run single transaction refresh to get some reference output.
-# The result of a query on the aggregate should always look like this example.
-permutation "R1_refresh" "R1_commit" "S1_select" "R2_commit" "R3_commit" "R4_commit"
+####################################################################
+#
+# Tests for concurrent updates to the invalidation threshold (first
+# transaction of a refresh).
+#
+####################################################################
+
+# Run single transaction refresh to get some reference output.  The
+# result of a query on the aggregate should always look like this
+# example (when refreshed with the same window).
+permutation "R1_refresh" "S1_select" "R3_refresh" "S1_select"  "L2_read_unlock_threshold_table" "L3_unlock_cagg_table" "L1_unlock_threshold_table"
+
+# A threshold reader (insert) should block a refresh if the threshold
+# does not exist yet (insert of new threshold)
+permutation "L2_read_lock_threshold_table" "R3_refresh" "L2_read_unlock_threshold_table" "S1_select" "L3_unlock_cagg_table" "L1_unlock_threshold_table"
+
+# A threshold reader (insert) should block a refresh if the threshold
+# needs an update
+permutation "R1_refresh" "L2_read_lock_threshold_table" "R3_refresh" "L2_read_unlock_threshold_table" "S1_select" "L3_unlock_cagg_table" "L1_unlock_threshold_table"
+
+# A threshold reader (insert) won't block a refresh if the threshold
+# doesn't need an update
+permutation "R3_refresh" "L2_read_lock_threshold_table" "R1_refresh" "L2_read_unlock_threshold_table"  "S1_select" "L3_unlock_cagg_table" "L1_unlock_threshold_table"
+
+#
+# The following two permutations test different paths for concurrent
+# inserts of new thresholds when one doesn't previously exist. However
+# the tests are flaky because the two refreshes block on a lock and
+# will then race to the a second lock when the first lock is released.
+#
+
+# Check that two refreshes (R1 then R3), which both try to insert a
+# new threshold, will not lead to a conflict. R1 and R3 should both
+# block on trying to insert a new threshold, but only one of them
+# should succeed and the other should do an update instead.
+#permutation "L1_lock_threshold_table" "R1_refresh" "R3_refresh" "L1_unlock_threshold_table" "L2_read_unlock_threshold_table" "L3_unlock_cagg_table" "S1_select"
```

Comment:
The two commented permutations here tests two concurrent inserts of a new threshold, but they are flaky since they rely on two blocked refreshes that race to insert a threshold after the lock is released. Enabling them leads to better test coverage but will make tests flaky. Continuing investigation, but figured best to leave in here for now given the expected questions on test coverage for the concurrent threshold insert case.